packages/sanity/src/core/canvas/actions/LinkToCanvas/useLinkToCanvas.ts

```diff
@@ -0,0 +1,191 @@
+import {type SanityClient, type SanityDocument} from '@sanity/client'
+import {useMemo, useState} from 'react'
+import {useObservable} from 'react-rx'
+import {catchError, combineLatest, map, type Observable, of, tap} from 'rxjs'
+
+import {createAppIdCache} from '../../../create/studio-app/appIdCache'
+import {useStudioAppIdStore} from '../../../create/studio-app/useStudioAppIdStore'
+import {useClient} from '../../../hooks/useClient'
+import {useProjectStore} from '../../../store/_legacy/datastores'
+import {type ProjectStore} from '../../../store/_legacy/project/types'
+import {useSource} from '../../../studio/source'
+import {useWorkspace} from '../../../studio/workspace'
+
+interface CanvasResponse {
+  error?: string
+  originalDocument?: SanityDocument
+  mappedDocument?: SanityDocument
+  canvasContent?: unknown
+  diff?: {
+    indexedPath: string[]
+    prevValue: unknown
+    value: unknown
+    type: string
+  }[]
+}
+
+type StudioToCanvasRequestBody = ({documentId: string} | {document: SanityDocument}) & {
+  projectUser?: {name: string}
+  linkedProjectTitle?: string
+  schemaId: string
+  localeSettings: {
+    timeZone: string
+    locale: string
+  }
+}
+const useWorkspaceSchemaId = () => {
+  const workspace = useWorkspace()
+  return `sanity.workspace.schema.${workspace.name || 'default'}`
+}
+interface UseLinkToCanvasResponse {
+  status: 'validating' | 'redirecting' | 'error' | 'missing-document-id' | 'diff'
+  error?: string | null
+  redirectUrl?: string
+  response?: CanvasResponse
+}
+const initialState: UseLinkToCanvasResponse = {
+  status: 'validating',
+}
+
+const CANVAS_CLIENT_CONFIG = {
+  apiVersion: 'v2025-04-29',
+}
+
+const getCanvasLinkUrl = ({
+  documentId,
+  workspaceName,
+  documentType,
+  applicationId,
+  client,
+  projectStore,
+}: {
+  documentId: string
+  workspaceName: string
+  documentType: string
+  applicationId: string
+  client: SanityClient
+  projectStore: ProjectStore
+}) => {
+  const dataset = client.config().dataset || ''
+  const projectId = client.config().projectId || ''
+  // TODO: If comlink is available don't get the org id, use comlink to navigate to canvas
+
+  return projectStore.getOrganizationId().pipe(
+    map((organizationId) => {
+      const queryParams = new URLSearchParams({
+        projectId,
+        dataset,
+        documentType,
+        documentId,
+        workspaceName,
+        applicationId,
+      })
+      const isStaging = client.config().apiHost === 'https://api.sanity.work'
+
+      return `https://www.sanity.${isStaging ? 'work' : 'io'}/@${organizationId}/canvas/studio-import?${queryParams.toString()}`
+    }),
+  )
+}
+
+const localSettings = Intl.DateTimeFormat().resolvedOptions()
```

Comment:
Nit: unsure about file global const conventions in studio, but I'd see this move to above all functions in the file.